mtags/src/main/scala/scala/meta/internal/pc/SourceModificator.scala

```diff
@@ -0,0 +1,104 @@
+package scala.meta.internal.pc
+
+import scala.meta.pc.OffsetParams
+import scala.meta.pc.DefinitionResult
+import scala.meta.internal.metals.CompilerOffsetParams
+import org.eclipse.lsp4j.TextEdit
+import org.eclipse.lsp4j.Hover
+import org.eclipse.lsp4j.CompletionList
+import org.eclipse.lsp4j.Position
+import org.eclipse.lsp4j.Range
+import scala.collection.JavaConverters._
+
+trait SourceModificator {
+  def forDefinition(pos: OffsetParams)(
+      f: OffsetParams => DefinitionResult
+  ): DefinitionResult
+  def forHover(pos: OffsetParams)(
+      f: OffsetParams => Option[Hover]
+  ): Option[Hover]
+  def forCompletion(pos: OffsetParams)(
+      f: OffsetParams => CompletionList
+  ): CompletionList
+}
+
+object SourceModificator {
```

Comment:
This looks good, but maybe we can do it alongside the Ammonite fixes in `Compilers.ammoniteInputPosOpt` and `Ammonite.adjustHoverResp` etc.

We wouldn't need to forward the auto imports lists and we could generalize the Ammonite approach.